File: server/services/battle/army.dart
import 'dart:math';
import 'package:models/models.dart';

import 'battle.dart';

enum ESide { OFF, DEF }

class Army {
  Army({
    required this.side,
    required this.population,
    required this.units,
    required this.numbers,
    this.party = false,
    this.brew = false,
    this.upgrades = const [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    this.mission = Mission.raid,
    this.targets = const [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  });

  final ESide side;
  final int population;
  final List<Unit> units;
  List<int> numbers;
  final List<int> upgrades;
  final Mission mission;
  final List<int> targets;
  final bool party;
  final bool brew;

  bool isScan() {
    var scanPosition = 0;
    for (var i = 0; i < units.length; i++) {
      if (units[i].unitKind == UnitKind.SPY) {
        scanPosition = i;
        break;
      }
    }
    return numbers[scanPosition] > 0;
  }

  BattlePoints getOff() {
    final result = zipWith(
      (unit, number, upgrade) {
        final points = number * upgradeMethod(unit.offense, upgrade);
        return BattlePoints.off(points, unit.infantry);
      },
      units,
      numbers,
      upgrades,
    );
    return result.fold(BattlePoints.zero(), (acc, val) => acc.add(val));
  }

  BattlePoints getDef() {
    final result = zipWith(
      (unit, number, upgrade) {
        final infantryPoints = upgradeMethod(unit.defenseInfantry, upgrade);
        final cavalryPoints = upgradeMethod(unit.defenseCavalry, upgrade);
        return BattlePoints(infantryPoints, cavalryPoints).mul(number);
      },
      units,
      numbers,
      upgrades,
    );
    return result.fold(BattlePoints.zero(), (acc, val) => acc.add(val));
  }

  double getScan() {
    final result = zipWith(
      (unit, number, upgrade) {
        return unit.unitKind == UnitKind.SPY
            ? number * upgradeMethod(unit.getS(), upgrade)
            : 0;
      },
      units,
      numbers,
      upgrades,
    );
    return result.fold(0, (acc, val) => acc + val);
  }

  double getScanDef() {
    final result = zipWith(
      (unit, number, upgrade) {
        return unit.unitKind == UnitKind.SPY
            ? number * upgradeMethod(unit.getSD(), upgrade)
            : 0;
      },
      units,
      numbers,
      upgrades,
    );
    return result.fold(0, (acc, val) => acc + val);
  }

  int getTotal() {
    return numbers.fold(0, (acc, val) => acc + val);
  }

  double upgradeMethod(int stat, int level) {
    final number = pow(1.015, level) * stat;
    return (1e-4 * number / 1e-4).round().toDouble();
  }

  static List<R> zipWith<A, B, C, R>(
    R Function(A, B, C) f,
    List<A> units,
    List<B> numbers,
    List<C> upgrades,
  ) {
    final result = <R>[];
    for (var i = 0; i < units.length; i++) {
      result.add(f(units[i], numbers[i], upgrades[i]));
    }
    return result;
  }

  void applyLosses(double losses, BattleResult battleResult) {
    final unitsBefore = <int>[];
    final casualties = <int>[];
    numbers = numbers.map((n) {
      unitsBefore.add(n);
      final after = (n * (1 - losses)).round();
      casualties.add(n - after);
      return after;
    }).toList();
    battleResult.unitsBeforeBattle?.add(unitsBefore);
    battleResult.casualties?.add(casualties);
  }

  List<int> rams() {
    for (var i = 0; i < 10; i++) {
      if (units[i].unitKind == UnitKind.RAM) {
        return [numbers[i], upgrades[i]];
      }
    }
    return [0, 0];
  }

  List<int> cats() {
    for (var i = 0; i < 10; i++) {
      if (units[i].unitKind == UnitKind.CAT) {
        return [numbers[i], upgrades[i]];
      }
    }
    return [0, 0];
  }
}

File: java_server/src/main/java/xyz/qruto/java_server/models/battle/Army.java
package xyz.qruto.java_server.models.battle;

import lombok.Builder;
import lombok.Data;
import xyz.qruto.java_server.models.Mission;
import xyz.qruto.java_server.models.units.Unit;
import xyz.qruto.java_server.models.units.UnitKind;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Data
@Builder
public class Army {
    private ESide side;
    private int population;
    private List<Unit> units;
    private List<Integer> numbers;
    @Builder.Default
    private List<Integer> upgrades = new ArrayList<>(List.of(0,0,0,0,0,0,0,0,0,0));
    @Builder.Default
    private Mission mission = Mission.attack;
    @Builder.Default
    private List<Integer> targets = new ArrayList<>(List.of(0,0,0,0,0,0,0,0,0,0));
    private boolean party;
    private boolean brew;

    public enum ESide{
        OFF, DEF
    }

    public boolean isScan() {
        int scanPosition = 0;
        for (int i = 0; i < units.size(); i++){
            if (units.get(i).getUnitKind().equals(UnitKind.SPY)){
                scanPosition = i;
                break;
            }
        }
        return numbers.get(scanPosition) > 0;
    }

    public BattlePoints getOff() {
        var result = Army.zipWith((unit, number, upgrade) -> {
            var points = number * upgrade(unit.getOffense(), upgrade);
            return BattlePoints.off(points, unit.isInfantry());
        }, units, numbers, upgrades);
        return result.stream().reduce(BattlePoints.zero(), BattlePoints::add);
    }

    public BattlePoints getDef() {
        var result = Army.zipWith(
                (unit, number, upgrade) ->
                        new BattlePoints(
                                upgrade(unit.getDefenseInfantry(), upgrade), upgrade(unit.getDefenseCavalry(), upgrade))
                                .mul(number), units, numbers, upgrades);
        return result.stream().reduce(BattlePoints.zero(), BattlePoints::add);
    }

    public double getScan() {
        var result = Army.zipWith(
                (unit, number, upgrade) ->
                        unit.getUnitKind().equals(UnitKind.SPY) ? number * upgrade(unit.getS(), upgrade) : 0
        , units, numbers, upgrades);
        return result.stream().reduce(0.0, Double::sum);
    }

    public double getScanDef() {
        var result = Army.zipWith(
                (unit, number, upgrade) ->
                        unit.getUnitKind().equals(UnitKind.SPY) ? number * upgrade(unit.getSD(), upgrade) : 0
                , units, numbers, upgrades);
        return result.stream().reduce(0.0, Double::sum);
    }

    public Integer getTotal() {
        return numbers.stream().reduce(0, Integer::sum);
    }

    private double upgrade(int stat, int level){
        double number = Math.pow(1.015, level) * stat;
        return (int) (1e-4 * Math.round(number / 1e-4));
    }

    public static <A, B, C, R> List<R> zipWith(ZipWith<A, B, C, R> f,
                      List<A> units,
                      List<B> numbers,
                      List<C> upgrades){
        var result = new ArrayList<R>();
        for (int i = 0; i < units.size(); i++){
            result.add(f.apply(units.get(i), numbers.get(i), upgrades.get(i)));
        }
        return result;
    }

    public void applyLosses(double losses, BattleResult battleResult) {
        List<Integer> unitsBefore = new ArrayList<>();
        List<Integer> casualties = new ArrayList<>();
        numbers = numbers.stream()
                .map(n -> {
                    unitsBefore.add(n);
                    var after = (int) Math.round(n * (1 - losses));
                    casualties.add(n - after);
                    return after;
                })
                .collect(Collectors.toList());
        battleResult.getUnitsBeforeBattle().add(unitsBefore);
        battleResult.getCasualties().add(casualties);
    }

    public int[] rams() {
        for (int i = 0; i < 10; i++) {
            if (units.get(i).getUnitKind().equals(UnitKind.RAM)) {
                return new int[]{numbers.get(i), upgrades.get(i)};
            }
        }
        return new int[]{0, 0};
    }

    public int[] cats() {
        for (int i = 0; i < 10; i++) {
            if (units.get(i).getUnitKind().equals(UnitKind.CAT)) {
                return new int[]{numbers.get(i), upgrades.get(i)};
            }
        }
        return new int[]{0, 0};
    }
}

